src/Features/CSharp/Portable/GenerateType/CSharpGenerateTypeService.cs

```diff
@@ -865,28 +874,58 @@ internal override bool IsSimpleName(ExpressionSyntax expression)
         }
 
         private ITypeSymbol GetPropertyType(
-            SimpleNameSyntax property,
+            SyntaxNode propertyNode,
             SemanticModel semanticModel,
             ITypeInferenceService typeInference,
             CancellationToken cancellationToken)
         {
-            var parent = property.Parent as AssignmentExpressionSyntax;
-            if (parent != null)
+            var parentAssignment = propertyNode.Parent as AssignmentExpressionSyntax;
+            if (parentAssignment != null)
             {
-                return typeInference.InferType(semanticModel, parent.Left, true, cancellationToken);
+                return typeInference.InferType(semanticModel, parentAssignment.Left, true, cancellationToken);
+            }
+
+            var patternNode = propertyNode as SubPropertyPatternSyntax;
+            if (patternNode != null)
+            {
+                return GetPatternType(patternNode.Pattern, semanticModel, cancellationToken);
             }
 
             return null;
         }
 
-        private IPropertySymbol CreatePropertySymbol(SimpleNameSyntax propertyName, ITypeSymbol propertyType)
+        private ITypeSymbol GetPatternType(
+            PatternSyntax pattern, SemanticModel semanticModel, CancellationToken cancellationToken)
         {
+            var typeInfo = pattern.TypeSwitch(
+                (DeclarationPatternSyntax declarationPattern) => semanticModel.GetTypeInfo(declarationPattern.Type),
+                (ConstantPatternSyntax constantPattern) => semanticModel.GetTypeInfo(constantPattern.Expression),
+                (RecursivePatternSyntax recursivePattern) => semanticModel.GetTypeInfo(recursivePattern.Type),
+                (PropertyPatternSyntax propertyPattern) => semanticModel.GetTypeInfo(propertyPattern.Type));
```

Comment:
i found it odd i needed to write this.  "GetTypeInfo" on a pattern returned nothing.  my *intuition* would be that it could report a type reasonably.  for example, if i have ```foo is int i``` and i ask for the type of ```int i```, then i'd get the type ```int``` back.  Similarly, if i had  ```a is B { X is 4 }```, then GetTypeInfo on the ConstantPattern for ```4``` would return ```int```.

However, if you don't want GetSymbolInfo working on Patterns, that's fine with me.  We'll just have to take that into account with IDE features like this.